### PR TITLE
Spotify Artist Search

### PR DESCRIPTION
**Pull request by hedgehog1833**
- It is now possible to search artists via the spotify rest api
- This is PR is work in progress 
- Tests are still missing
- Some todos in code
- Caching of auth token still missing

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[Spotify Search](https://trello.com/c/lIsyI6P8/189-spotify-search)

**Comment by hedgehog1833** [comment]
- Master merged
- Following now also works on Spotfiy
- Tests still missing, coming next 💦

**Comment by NilsDammenhayn** [comment]
I proudly present: the Spotify search :)
Have fun with this gigantic PR 😂

File: webapp/src/main/java/rocks/metaldetector/service/artist/ArtistsServiceImpl.java [hunk 1]
```diff
@@ -59,4 +62,17 @@ public DiscogsArtistSearchResultDto searchDiscogsByName(String artistQueryString
...
     return result;
   }
+
+  @Override
+  @Transactional
+  public SpotifyArtistSearchResultDto searchSpotifyByName(String artistQueryString, Pageable pageable) {
+    String publicUserId = currentPublicUserIdSupplier.get();
+    UserEntity userEntity = userRepository.findByPublicId(publicUserId).orElseThrow(() ->
+            new ResourceNotFoundException("User with public id '" + publicUserId + "' not found!")
+    );
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/artist/ArtistsServiceImpl.java:72]: Please move this into separate method, it's a duplicate.

File: webapp/src/main/java/rocks/metaldetector/web/api/response/ArtistSearchResponseEntryDto.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+@AllArgsConstructor
+@Builder
+public class ArtistSearchResponseEntryDto {
+
+  private String id;
+  private String name;
+  private String imageUrl;
+  private String uri;
+  private String source;
+  private boolean followed;
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/web/api/response/ArtistSearchResponseEntryDto.java:19]: So, now we have our own Dto in web module you can remove the `followed` field from Discogs and Spotify modules. The Spotify and Discogs module can't know if some of our users follow the requested artist.

The value for `followed` should be set in `ArtistServiceImpl`.

File: discogs/src/test/java/rocks/metaldetector/discogs/client/transformer/DiscogsArtistSearchResultContainerTransformerTest.java [hunk 3]
```diff
@@ -61,4 +63,32 @@ void should_transform() {
...
+    // then
+    assertThat(result.getSearchResults().size()).isEqualTo(3);
+  }
+
+  @Test
+  @DisplayName("Should cut suffix ' ($number)' from artist name")
+  void should_transform_artist_name() {
+    // given
+    var artistName = "Karg";
+    DiscogsArtistSearchResultContainer container = DiscogsArtistSearchResultContainerFactory.withArtistNames(List.of("Karg (3)"));
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply discogs/src/test/java/rocks/metaldetector/discogs/client/transformer/DiscogsArtistSearchResultContainerTransformerTest.java:85]: No it doesn't. And I don't care, super edge case.

File: spotify/src/main/java/rocks/metaldetector/spotify/client/SpotifyAuthenticationClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  static final String AUTHORIZATION_HEADER_PREFIX = "Basic ";
+
+  private final RestTemplate spotifyRestTemplate;
+  private final SpotifyConfig spotifyConfig;
+
+  @Override
+  public String getAuthenticationToken() {
+    HttpEntity<MultiValueMap<String, String>> httpEntity = createHttpEntity();
+    ResponseEntity<SpotifyAuthenticationResponse> responseEntity = spotifyRestTemplate.postForEntity(
+        spotifyConfig.getAuthenticationBaseUrl() + AUTHORIZATION_ENDPOINT, httpEntity, SpotifyAuthenticationResponse.class);
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply spotify/src/main/java/rocks/metaldetector/spotify/client/SpotifyAuthenticationClientImpl.java:42]: I think putting the `RestClientException` in the ExceptionHandler is cleaner as it then also works correctly for the discogs module.

File: spotify/src/main/java/rocks/metaldetector/spotify/api/search/SpotifyArtistSearchResultContainer.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+@NoArgsConstructor
+@Builder
+@JsonPropertyOrder({
+    "artists"
+})
+public class SpotifyArtistSearchResultContainer {
+
+  private SpotifyArtistSearchResult artists;
+
+}
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply spotify/src/main/java/rocks/metaldetector/spotify/api/search/SpotifyArtistSearchResultContainer.java:20]: As discussed: yes, we do..
